package idgs.pb;

option java_outer_classname = "PbRpcMessage";

// option optimize_for = CODE_SIZE;
option optimize_for = SPEED;

enum MemberIdConst {
  ALL_MEMBERS    = -1;
  ANY_MEMBER     = -2;
  UNKNOWN_MEMBER = -3;
  CLIENT_MEMBER  = -4;
}

enum TransportChannel {
  TC_AUTO       = 0;  // Selected by actor framework, TCP for size > MTU, UNICAST for size < MTU
  TC_TCP        = 1;  // TCP, for large message
  TC_UNICAST    = 2;  // unicast UDP
  TC_MULTICAST  = 3;  // multicast UDP
}

message ActorId {
  required int32 member_id = 1 [default = -3];
  required string actor_id = 2 ;
}

// how payload and attachement serdes
enum PayloadSerdes {
  PB_BINARY     = 0; // protobuf binary 
  PB_JSON       = 1; // json
  PB_TEXT       = 2; // protobuf text
}

message RpcMessage {
  required string operation_name = 1;

  required ActorId source_actor = 2;
  required ActorId dest_actor = 3;
  
  optional uint64 message_id = 4;
  
  optional TransportChannel channel = 5 [default = TC_AUTO];

  optional ActorId client_actor = 6;
  
  optional PayloadSerdes serdes_type = 7 [default = PB_BINARY];

  // payload of this message, type is specified by operationType
  optional bytes payload = 8;
  
  message NameValuePair {
    required string name = 1;
    required bytes value = 2;
  };
  repeated NameValuePair attachments = 9;  
}


//
// actor operation descriptor
//
message ActorOperationDescripor {
  required string name = 1;
  optional string description = 2;
  optional string payload_type = 3;
  repeated string response_payload_type = 4;
}

enum ActorType {
  AT_STATELESS  = 0; // stateless actor
  AT_STATEFUL   = 1; // stateful actor
}

//
// actor descriptor
//
message ActorDescriptor {
  required string name = 1;
  optional ActorType type = 2 [default = AT_STATEFUL ];
  optional string description = 3;
  
  repeated ActorOperationDescripor in_operations = 4;
  repeated ActorOperationDescripor out_operations = 5;
  repeated string consume_actor_name = 6;
}

//
// module descriptor
//
message ModuleDescriptor {
  required string name = 1;
  optional string description = 2;
  repeated ActorDescriptor actors = 3;
}